### PR TITLE
Refactor protocol proxy to support ExtJS 4 and 5.

### PR DESCRIPTION
**Pull request by marcjansen**
With this PR in, all tests pass for me in Chrome 33 and Firefox 34.

### CODE REVIEW

File: src/GeoExt/data/proxy/Protocol.js [hunk 1]
```diff
@@ -22,27 +22,67 @@ Ext.define('GeoExt.data.proxy.Protocol', {
...
+
+    },
 
     /**
-     * Should options.params be set directly on options before passing it into
-     * the protocol's read method?
+     * We need to override this as the OpenLayers classes passed as configs
+     * loose their class-nature and seem to be copied by ExtJS somewhere.
+     *
+     * We deal with this elsewhere in a different manor and should see if
```
Comment (bartvde) [inline src/GeoExt/data/proxy/Protocol.js]: manor -> manner

File: src/GeoExt/data/proxy/Protocol.js [hunk 2]
```diff
@@ -22,27 +22,67 @@ Ext.define('GeoExt.data.proxy.Protocol', {
...
+     * we can either share code or get rid of this special handling all
+     * together. The problem isn't reproducible for other 'classes' with a
+     * similar inheritance startegy as OpenLayers 2 has.
```
Comment (bartvde) [inline src/GeoExt/data/proxy/Protocol.js]: startegy -> strategy

File: src/GeoExt/data/proxy/Protocol.js [hunk 3]
```diff
@@ -22,27 +22,67 @@ Ext.define('GeoExt.data.proxy.Protocol', {
...
      *
-     * @cfg {Boolean}
+     * TODO Find a way to have this functionality shared or get rid of it.
      */
-    setParamsAsOptions: false,
+    initConfig: function(config){
+        var me = this,
+            cfg = config || {},
+            prefix = me.$configPrefixed ? '_' : '',
```
Comment (bartvde) [inline src/GeoExt/data/proxy/Protocol.js]: can you explain more about this option?

**Comment by bartvde** [comment]
LGTM only minor comments

File: src/GeoExt/data/proxy/Protocol.js [hunk 4]
```diff
@@ -63,22 +103,28 @@ Ext.define('GeoExt.data.proxy.Protocol', {
...
 
         if(GeoExt.isExt5){
-            callback = operation.getCallback();
-            scope = operation.getScope();
-            params = Ext.applyIf(operation.getParams() || {}, me.extraParams || {});
-        }
-        else {
-            params = Ext.applyIf(operation.params || {}, me.extraParams || {});
+            callback = callback || operation.getCallback();
+            scope = scope || operation.getScope();
```
Comment (chrismayer) [inline src/GeoExt/data/proxy/Protocol.js]: I guess this is done because in ExtJS5 there are no arguments ``callback`` and ``scope``?

**Comment by chrismayer** [comment]
Wow, thank you guys! :clap:  One of the last big things so far...

Tests for this pass in IE 9, FF34 and Chrome 39. One minor question, see above.

File: src/GeoExt/data/proxy/Protocol.js (hunk 3 above)
Comment (marcjansen) [inline src/GeoExt/data/proxy/Protocol.js]: I don't know if it is standard, I found out about it via reverse engineering. As this whole method is more or less a workaround and it works for both versions, I'd leave it in.

Before the extjs5 branch is ready for prime time, every change will be re-reviewed for sure.

File: src/GeoExt/data/proxy/Protocol.js (hunk 4 above)
Comment (marcjansen) [inline src/GeoExt/data/proxy/Protocol.js]: Good question. I am not 100% sure to be honest. In my current understanding I'd say that we should have had something similar in ExtJS 4 all the time. This is certainly a thing we should look at once we start rereviewing the branch in total.

**Comment by marcjansen** [comment]
I'll merge this now; thanks for the reviews.

Is I said in the comments to the questions, I think we should actually re-review any change of the extjs5 branch before we eventually merge it and make it the new `master`.

The whole thing has been cooking for so long now, I want a green state to revisit any changes it introduces. Certain problems are addressed differently throughout the various commits. I would love to have this a little bit more harmonized.

**Pull request merged**